### DISCUSSION TITLE
Creating a user programmatically using Beanie

### DISCUSSION THREAD

**Discussion by jolmedo** [fastapi-users/fastapi-users#1100] [Q&A] [answered]
I was reading the Fastapi-users official documents at (https://fastapi-users.github.io/), trying to find a way to create a user programmatically. In fact, there's an example of it, available at [here](https://fastapi-users.github.io/fastapi-users/10.2/cookbook/create-user-programmatically/) using SQLAlchemy not Beanie. How can I do it using Beanie? Do you have that procedure in one of the documents? Of course, they are not the same thing: I'm not pretty sure, but I think SQLAlchemy is synchronous, am I right? Beanie is asynchronous by nature, so there's no need to call an inexistent get_async_session function. Issue is there isn't any topic to explain how to do it. Can you please give us more information about this? It would be nice to create a user programmatically using Beanie. Maybe it's obvious, but I can't see it. Thanks in advance.

**Comment by frankie567** [comment, marked as answer]
FastAPI Users fully embrace asyncio, so it's both async for SQLAlchemy and Beanie.

As for creating a user programmatically with Beanie, the mechanism is fully similar. The key thing to understand is that we need to instantiate all the dependencies we need.

It's true that, for Beanie, there is no `get_async_session`. But we still need `get_user_db` and `get_user_manager`. So it would look like this:

```py
import contextlib

from app.db import get_user_db
from app.schemas import UserCreate
from app.users import get_user_manager
from fastapi_users.exceptions import UserAlreadyExists

get_user_db_context = contextlib.asynccontextmanager(get_user_db)
get_user_manager_context = contextlib.asynccontextmanager(get_user_manager)


async def create_user(email: str, password: str, is_superuser: bool = False):
    try:
        async with get_user_db_context() as user_db:
            async with get_user_manager_context(user_db) as user_manager:
                user = await user_manager.create(
                    UserCreate(
                        email=email, password=password, is_superuser=is_superuser
                    )
                )
                print(f"User created {user}")
    except UserAlreadyExists:
        print(f"User {email} already exists")
```

**Comment by Dogrtt** [comment]
Hi, I think it's not necessary to create a new Question, so I would like to ask here.
I prepare example repo that illustrate my problem: https://github.com/Dogrtt/fastapi-users-beanie-pytest-failure
What I'm doing wrong?

**Reply by frankie567** [reply]
It's probably because Beanie isn't initialized when running your unit tests. The call we have here in FastAPI startup event: https://github.com/Dogrtt/fastapi-users-beanie-pytest-failure/blob/65c3b5fed822c6664428a48da329971e6e0f3468/main.py#L51-L56 is critical to have before any attempt to use Beanie. You can probably put this into an auto-use fixture.

**Reply by Dogrtt** [reply]
Got it, thank you.

**Reply by Dogrtt** [reply]
> It's probably because Beanie isn't initialized when running your unit tests. The call we have here in FastAPI startup event: https://github.com/Dogrtt/fastapi-users-beanie-pytest-failure/blob/65c3b5fed822c6664428a48da329971e6e0f3468/main.py#L51-L56 is critical to have before any attempt to use Beanie. You can probably put this into an auto-use fixture.

It did not help, but I took a look how tests organized in the Beanie's repo:
https://github.com/roman-right/beanie/blob/main/tests/fastapi/conftest.py

They are using https://github.com/florimondmanca/asgi-lifespan lib for programmatical ASGI app start.
That's working good.

**Comment by sblvkr** [comment]
I use the solution presented above, but I get the following error:
```
Traceback (most recent call last):
  File "E:\cms\src\helpers\create_superuser.py", line 28, in <module>
    asyncio.run(create_user("king.arthur@camelot.ua", "guinevere"))
  File "C:\Users\sobki\AppData\Local\Programs\Python\Python311\Lib\asyncio\runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\sobki\AppData\Local\Programs\Python\Python311\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\sobki\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "E:\cms\src\helpers\create_superuser.py", line 18, in create_user
    user = await user_manager.create(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\cms\.venv11\Lib\site-packages\fastapi_users\manager.py", line 131, in create
    existing_user = await self.user_db.get_by_email(user_create.email)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\cms\.venv11\Lib\site-packages\fastapi_users_db_beanie\__init__.py", line 75, in get_by_email
    self.user_model.email == email,
    ^^^^^^^^^^^^^^^^^^^^^
  File "E:\cms\.venv11\Lib\site-packages\pydantic\_internal\_model_construction.py", line 221, in __getattr__
    raise AttributeError(item)
AttributeError: email
```
In my schema and model there is an "email" field. What could be the catch?

**Comment by hgalytoby** [comment]
> I use the solution presented above, but I get the following error:
> 
> ```
> Traceback (most recent call last):
>   File "E:\cms\src\helpers\create_superuser.py", line 28, in <module>
>     asyncio.run(create_user("king.arthur@camelot.ua", "guinevere"))
>   File "C:\Users\sobki\AppData\Local\Programs\Python\Python311\Lib\asyncio\runners.py", line 190, in run
>     return runner.run(main)
>            ^^^^^^^^^^^^^^^^
>   File "C:\Users\sobki\AppData\Local\Programs\Python\Python311\Lib\asyncio\runners.py", line 118, in run
>     return self._loop.run_until_complete(task)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\sobki\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 653, in run_until_complete
>     return future.result()
>            ^^^^^^^^^^^^^^^
>   File "E:\cms\src\helpers\create_superuser.py", line 18, in create_user
>     user = await user_manager.create(
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "E:\cms\.venv11\Lib\site-packages\fastapi_users\manager.py", line 131, in create
>     existing_user = await self.user_db.get_by_email(user_create.email)
>                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "E:\cms\.venv11\Lib\site-packages\fastapi_users_db_beanie\__init__.py", line 75, in get_by_email
>     self.user_model.email == email,
>     ^^^^^^^^^^^^^^^^^^^^^
>   File "E:\cms\.venv11\Lib\site-packages\pydantic\_internal\_model_construction.py", line 221, in __getattr__
>     raise AttributeError(item)
> AttributeError: email
> ```
> 
> In my schema and model there is an "email" field. What could be the catch?

Did you run `init beanie`?

[init_beanie](https://fastapi-users.github.io/fastapi-users/latest/configuration/databases/beanie/#initialize-beanie)

https://beanie-odm.dev/tutorial/initialization/

**Reply by sblvkr** [reply]
Yep, my main.py
```
client = AsyncIOMotorClient(
    settings.MONGODB_URI,
    uuidRepresentation='standard',
)
@app.on_event("startup")
async def start_db():
    await init_beanie(
        database=getattr(client, settings.MONGODB_DB_NAME),
        document_models=[Cluster, Database, Message, User, Role],
    )
```
